plugins/org.python.pydev.parser/src/org/python/pydev/parser/prettyprinterv2/PrettyPrinterVisitorV2.java

```diff
@@ -122,18 +122,36 @@ public Object visitAssign(Assign node) throws Exception {
         java.util.List<ILinePart> recordChanges = null;
         org.python.pydev.shared_core.structure.Tuple<ILinePart, ILinePart> lowerAndHigher = null;
 
-        for (int i = 0; i < node.targets.length; i++) {
-            exprType target = node.targets[i];
-            if (i >= 1) { //more than one assign
+        if (node.targets.length > 0) {
+            for (int i = 0; i < node.targets.length; i++) {
+                exprType target = node.targets[i];
+                if (i >= 1) { //more than one assign
+                    doc.add(lowerAndHigher.o2.getLine(), lowerAndHigher.o2.getBeginCol(),
+                            this.prefs.getAssignPunctuation(), node);
+                }
+                id = this.doc.pushRecordChanges();
+                target.accept(this);
+                recordChanges = this.doc.popRecordChanges(id);
+                lowerAndHigher = doc.getLowerAndHigerFound(recordChanges);
+            }
+            if (node.type != null) {
                 doc.add(lowerAndHigher.o2.getLine(), lowerAndHigher.o2.getBeginCol(),
-                        this.prefs.getAssignPunctuation(), node);
+                        this.prefs.getTypePunctuationColon(), node);
+                id = this.doc.pushRecordChanges();
+                node.type.accept(this);
+                recordChanges = this.doc.popRecordChanges(id);
+                lowerAndHigher = doc.getLowerAndHigerFound(recordChanges);
             }
+            doc.add(lowerAndHigher.o2.getLine(), lowerAndHigher.o2.getBeginCol(), this.prefs.getAssignPunctuation(),
+                    node);
+        } else if (node.type != null) {
```

Comment:
I don't think we need this `else` case at all (I don't think that in practice we can even reproduce a use case where `node.targets.length == 0` and `node.type != null`).

Also, I don't think we should special-case the `if(node.targets.length) > 0` at all (i.e.: always iterate through the targets should be fine even if its length == 0).

Reply:
When the statement is just something like this: `var: type`, node.targets comes empty. But not only that, instead of the type being stored in node.type, it is stored in the node.value. So it is a whole new case.

In short, node.type stores different types of values depending on whether node.targets is empty or not

Reply:
This is a bug in the AST generation then (as such, it should be fixed there).

We should do this in this pull request too.

i.e.: Create a test with that structure in: `org.python.pydev.parser.PyParser38Test` -- the actual fix is probably in `org.python.pydev.parser.grammar38.TreeBuilder38` when dealing with `JJTANN_ASSIGN` -- it should deal with the case where the value is not given (check the `arity` variable for that) -- and we have to backport to the 3.6 grammar tree builders as well.

Reply:
Note: it's still missing backporting the `JJTANN_ASSIGN` to `TreeBuilder36`... You can probably put that handling in some method in the `AbstractTreeBuilder` and then call it in both `TreeBuilder36` and `TreeBuilder38`.

Reply:
Ok. I forgot to backport it